### PR TITLE
Add some integration tests.

### PR DESCRIPTION
**Pull request by mo-esmp**

### CODE REVIEW

File: TodoWithControllers/Program.cs [hunk 1]
```diff
@@ -1,24 +1,26 @@
...
-        static async Task Main(string[] args)
+        private static async Task Main(string[] args)
         {
             var builder = WebApplication.CreateBuilder(args);
 
             builder.Services.AddDbContext<TodoDbContext>(options => options.UseInMemoryDatabase("Todos"));
             builder.Services.AddControllers();
 
             var app = builder.Build();
+            if (args?.Length > 0)
```
Comment (davidfowl) [inline TodoWithControllers/Program.cs:18]: Ugh, no 😄

File: TodoWithControllers/Program.cs [hunk 2]
```diff
@@ -1,24 +1,26 @@
...
 using Microsoft.AspNetCore.Builder;
 using Microsoft.EntityFrameworkCore;
 using Microsoft.Extensions.DependencyInjection;
+using System.Threading.Tasks;
 
-namespace Todos
+namespace TodoWithControllers
 {
-    class Program
+    public class Program
```
Comment (davidfowl) [inline TodoWithControllers/Program.cs:8]: Also no.

File: TodoWithControllers/Program.cs [hunk 3]
```diff
@@ -1,24 +1,26 @@
...
     {
-        static async Task Main(string[] args)
+        private static async Task Main(string[] args)
```
Comment (davidfowl) [inline TodoWithControllers/Program.cs:10]: Nope.

**Comment by Carl-Hugo** [comment]
Why not create an abstract base class with four tests, then inherit from it loading a different program? 

That would:

1. Be way easier to maintain
2. Test the correctness of all APIs. They should enforce the same contracts as they share the same features but using different language capabilities.

---

Idea 2:  we could run a Postman collection as a GitHub action using Newman (I guess someone already created that action).

**Comment by davidfowl** [comment]
Please make a new PR that doesn't change all of the formatting.

File: TodoWithControllers/Program.cs (hunk 1 above)
Comment (mo-esmp) [reply TodoWithControllers/Program.cs:18]: Have no idea send request to what address when run `Program.cs` via reflection.
```csharp
var methodInfo = type.GetMethod("Main", BindingFlags.NonPublic | BindingFlags.Static);
const string port = "5001";
string[] args = { port };
methodInfo.Invoke(null, new[] { args });
```
Or something like this:

```csharp
var directoryInfo = Directory.GetParent(Directory.GetCurrentDirectory());
var solutionDirectory = $"{directoryInfo.Parent.Parent.Parent.Parent.FullName}";
var projectName = type.FullName.Split('.')[0];
var launchSettingsPath = $"{solutionDirectory}\\{projectName}\\Properties\\launchSettings.json";

using var file = File.OpenText(launchSettingsPath);
using var reader = new JsonTextReader(file);
dynamic o2 = JToken.ReadFrom(reader);

var port = o2.iisSettings.iisExpress.sslPort;
Client = new HttpClient { BaseAddress = new Uri($"https://localhost:{port}") };
```

File: TodoWithControllers/Program.cs (hunk 2 above)
Comment (mo-esmp) [reply TodoWithControllers/Program.cs:8]: Then I have to create test project for each API project, it's OK 😕 .

File: TodoWithControllers/Program.cs (hunk 3 above)
Comment (mo-esmp) [reply TodoWithControllers/Program.cs:10]: That's Resharper fault, not me 😄.

**Comment by mo-esmp** [comment]
Testing code is hard (at least for me) without adding some level of abstraction. Perhaps you could guide me on this case.

File: TodoWithControllers/Program.cs (hunk 1 above)
Comment (Carl-Hugo) [reply TodoWithControllers/Program.cs:18]: Tip of the day:

- `ASPNETCORE_URLS`
- [Basic tests with the default WebApplicationFactory](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-3.1#basic-tests-with-the-default-webapplicationfactory)

Comment (mo-esmp) [reply TodoWithControllers/Program.cs:18]: @Carl-Hugo thanks for the tip. How about `--urls` command-line argument.
```csharp
string[] args = { "urls=https://localhost:5002" };
methodInfo.Invoke(null, new[] { args });
```

Comment (Carl-Hugo) [reply TodoWithControllers/Program.cs:18]: > @Carl-Hugo thanks for the tip. How about `--urls` command-line argument.
> 
> ```cs
> string[] args = { "urls=https://localhost:5002" };
> methodInfo.Invoke(null, new[] { args });
> ```

As far as I know, they do the same thing; `args` being the last "settings override".

**Pull request closed without merging**